File: 2020/day23/cups.ts
import { Node, pushValue } from "../../helpers/linked_list";

function day23(input: string) {
  console.time("Processing");
  // parse the data
  const cupsArray = input.split("").map((cup) => parseInt(cup, 10));
  const maxCup = 1000000;
  const minCup = 1;

  // quick lookup from label to cup
  const lookup = Array<Node<number>>(1000001);

  // add the fixed cups
  let head = new Node(cupsArray[0]);
  lookup[cupsArray[0]] = head;
  cupsArray.slice(1).forEach((cup) => {
    const node = pushValue(cup, head);
    lookup[cup] = node;
  });
  // add the remaining 1m cups
  for (let i = 10; i <= 1000000; i++) {
    const node = pushValue(i, head);
    lookup[i] = node;
  }
  // start the game
  let currentCup = head;
  for (let move = 0; move < 10000000; move++) {
    // cut out three cups to the right of the current cup
    const takenCups = currentCup!.right!;
    currentCup.right = currentCup?.right?.right?.right?.right;
    // where to put the cups?
    let searchValue = currentCup.value;
    let destination: Node<number> | undefined = undefined;
    while (destination === undefined) {
      // decrease search and wrap around
      searchValue = searchValue - 1;
      if (searchValue < minCup) {
        searchValue = maxCup;
      }
      // don't include the cups we've taken out
      if (
        searchValue !== takenCups.value &&
        searchValue !== takenCups.right!.value &&
        searchValue !== takenCups.right!.right!.value
      ) {
        destination = lookup[searchValue];
      }
    }
    // insert the cups back into the list
    takenCups.right!.right!.right = destination.right;
    destination.right = takenCups;
    // move to the next cup
    currentCup = currentCup.right!;
  }
  console.timeEnd("Processing");
  // part 2 results
  const one = lookup[1];
  console.log(
    "Done!",
    one.right!.value,
    one.right!.right!.value,
    one.right!.value * one.right!.right!.value
  );
}

day23("916438275");

File: 2020/day23/cups.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = self
        self.right = self


def push_value(value, head):
    new_node = Node(value)
    new_node.right = head
    new_node.left = head.left.right
    head.left.right = new_node
    head.left = new_node
    return new_node


def day23(input):
    # parse the data
    cups_array = [int(char) for char in input]
    max_cup = 1000000
    min_cup = 1

    # quick lookup from label to cup
    lookup = [None] * int(1000001)

    # add the fixed cups
    head = Node(cups_array[0])
    lookup[cups_array[0]] = head
    for cup in cups_array[1:]:
        node = push_value(cup, head)
        lookup[cup] = node

    # add the remaining 1m cups
    for i in range(10, 1000001):
        node = push_value(i, head)
        lookup[i] = node

    # start the game
    current_cup = head
    for _ in range(10000000):
        # cut out three cups to the right of the current cup
        taken_cups = current_cup.right
        current_cup.right = current_cup.right.right.right.right
        # where to put the cups?
        search_value = current_cup.value
        destination = None
        while destination is None:
            # decrease search and wrap around
            search_value = search_value - 1
            if search_value < min_cup:
                search_value = max_cup
            # don't include the cups we've taken out
            if (
                search_value != taken_cups.value
                and search_value != taken_cups.right.value
                and search_value != taken_cups.right.right.value
            ):
                destination = lookup[search_value]
        # insert the cups back into the list
        taken_cups.right.right.right = destination.right
        destination.right = taken_cups
        # move to the next cup
        current_cup = current_cup.right

    # part 2 results
    one = lookup[1]
    print(
        "Done!",
        one.right.value,
        one.right.right.value,
        one.right.value * one.right.right.value,
    )


day23("916438275")
